### ISSUE TITLE
Lighthouse (Chrome 98.0.4758.74) hangs on https://webvm.io

### ISSUE THREAD

**Issue by carlopi** [GoogleChrome/lighthouse#13613]
### FAQ

- [X] Yes, my issue is not about [variability](https://github.com/GoogleChrome/lighthouse/blob/master/docs/variability.md) or [throttling](https://github.com/GoogleChrome/lighthouse/blob/master/docs/throttling.md).
- [X] Yes, my issue is not about a specific accessibility audit (file with [axe-core](https://github.com/dequelabs/axe-core) instead).

### URL

https://webvm.io

### What happened?

Lighthouse hangs, it's 5 hours and counting.

### What did you expect?

Score computation to end at some point.

### What have you tried?

Replicated it.
Tried other versions of Chrome, on version 97.0.4692.99 it does complete but say NO_LCP (this might be related?)

### How were you running Lighthouse?

Chrome DevTools

### Lighthouse Version

Lighthouse 9.1.0

### Chrome Version

Chrome 98.0.4758.74

### OS

Linux

### Relevant log output

```shell
none
```

**Comment by connorjclark** [comment]
So the run gets a PROTOCOL_TIMEOUT for some reason. I didn't investigate why specifically, as the problem with the UI hanging indefinitely even after the run is finished (error'd or otherwise) is a big issue we need to fix. Some notes on that issue:

![image](https://user-images.githubusercontent.com/4071474/152033725-f39fc513-6db9-4fee-a1b1-d7dac0e27df7.png)

I traced that the lighthouse DT worker throws a fatal error, that is then caught and rethrown in `LighthousePanel.ts` by `startLighthouse` https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/panels/lighthouse/LighthousePanel.ts;l=354;drc=2d9a95ce7b9521ac5d724950938e21c5730ec343

But then it's just ignored https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/panels/lighthouse/LighthousePanel.ts;l=103;drc=2d9a95ce7b9521ac5d724950938e21c5730ec343

I'm having trouble telling how an error status is rendered in the existing code. Perhaps that has broken completely now. https://chromiumdash.appspot.com/commit/6c6dc45441d27e4912ce890fe3628a9d9e4bdd0d had fixed some part of the puzzle, but apparently not all of it.

______

To address your issue a bit, the NO_LCP is expected as there is no LCP candidates on the page, it being mostly canvas. atm canvas is not eligible to be LCP, although there is work being done to change that in Chromium. ref #12309

If you open DevTools-on-DevTools (use the keyboard shortcut to open DevTools on an undocked DevTools) you'll see the error being printed. Does it hang + print PROTOCOL_TIMEOUT every time for you?

**Comment by connorjclark** [comment]
> But then it's just ignored source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/panels/lighthouse/LighthousePanel.ts;l=103;drc=2d9a95ce7b9521ac5d724950938e21c5730ec343

That's not right, I overlooked a catch.

![image](https://user-images.githubusercontent.com/4071474/152039736-bd37ca0d-99da-4f7c-962d-5f924bf6d383.png)

it is hanging on the reset emulation call.

I got no hang on Chrome 100. can you try chrome canary?

Seems similar to #13396 but that should have been resolved in M97...

**Comment by carlopi** [comment]
Chrome 98.0.4758.80 (Official Build) beta (64-bit) I get this with DevTools-on-DevTools:
![Screenshot from 2022-02-01 21-19-59](https://user-images.githubusercontent.com/842657/152044972-b26d476b-ebaa-40c6-832b-4a6938171c5f.png)

**Comment by carlopi** [comment]
While on Chrome 99.0.4844.16 (Official Build) dev (64-bit), DevTools-on-DevTools I get:
![Screenshot from 2022-02-01 21-25-03](https://user-images.githubusercontent.com/842657/152045533-4d9f0930-19ad-4a36-aeeb-951c71a53420.png)

**Comment by carlopi** [comment]
I don't have a Canary installation, not sure what's the easiest way to get it on Ubuntu

**Comment by carlopi** [comment]
Meanwhile the package google-chrome-unstable has been updated to 100.0.4867.0-1, but I see the same log as for Chrome 99.0.4844.16 (I expanded it so more information are available):
![Screenshot from 2022-02-07 11-26-42](https://user-images.githubusercontent.com/842657/152770562-f7b447c0-7ea7-4f8f-a668-33e663ba459f.png)

**Comment by connorjclark** [comment]
That "Blackbox pattern" error sometimes happens and can be ignored, it's not related to Lighthouse.

I _think_ what's happening in your case is that you see no other logs because of a silly bug in the console where messages from different targets (in this case, Lighthouse running in a worker) are filtered out. The controls for the filter can be toggled:

![image](https://user-images.githubusercontent.com/4071474/153068803-97d56ace-acbc-4276-aef8-cd8570d35caf.png)

and make sure Selected Context option is turned off (it's a [footgun](https://bugs.chromium.org/p/chromium/issues/detail?id=1264583&q=&can=4)!) 
![image](https://user-images.githubusercontent.com/4071474/153068869-e5133765-076e-4883-af52-97437d04759c.png)

So I'm thinking that _maybe_ the Ligthouse logging messages were being hidden from you, and perhaps you didn't wait long enough (no more than 2-3 minutes) when trying in these other versions of Chrome?

> I don't have a Canary installation, not sure what's the easiest way to get it on Ubuntu

Not canary, but this is ToT (top of tree) so probably good enough: https://download-chromium.appspot.com/

_____

Perhaps this is an ubunutu-only bug.

**Comment by carlopi** [comment]
You are right, not all logging was enable, still what I see is after this:
![Screenshot from 2022-02-09 22-29-39](https://user-images.githubusercontent.com/842657/153293264-05dfbf2d-5b25-4920-93fa-cc598d20386b.png)
is a loop of setTimeout and currentPhase never be unset so the loop just keeps repeating:
![Screenshot from 2022-02-09 22-30-39](https://user-images.githubusercontent.com/842657/153293464-cdc618a2-db2d-44e0-b482-f1bb2d031454.png)

**Comment by connorjclark** [comment]
@carlopi just checking in, are you still finding that LH hangs on this site?

**Comment by connorjclark** [comment]
can still repro:

1. load https://webvm.io/
2. run Lighthouse report - specific settings don't seem to matter

This will hang after the Lighthouse worker returns the result. The UI will never close the loading dialog, because it hangs in `LighthouseController.cancel()`. A report generates just fine via the Node CLI.

control flow:

1. calls `LighthouseController.collect()` - LH runs, returns a result, and the `finally` block is reached
2. calls `LighthouseController.cancel()`
3. calls `this.protocolService.detach()` https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/panels/lighthouse/LighthouseController.ts;l=317;drc=bb058b71fa2283b99f807e7e211559ccf2781367
4. calls `resumeAllTargets()`
5. hangs on `Promise.all(resumePromises)` https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/core/sdk/TargetManager.ts;l=115;drc=fcc8e2e643f4beab78d87a7a8566f0ed50332666

From inspection, those promises map to these target names:

```
> [...this.#targets.values()].map(t => t.name())
[
    "Tab",
    "warmup.html",
    "Main",
    "edd98adc-6370-44b9-a637-8a550e35fe3a",
    "3da3432c-00a0-43ab-ba43-8a7b094cef27"
]
```

It's the last one that does not resolve. Its target info:

```
> [...this.#targets.values()].at(-1).targetInfo()
{
    "targetId": "CCD713886845AF410F6B3315656E9F79",
    "type": "worker",
    "title": "blob:https://webvm.io/53815c6a-4509-4c80-a9db-1c81410b4800",
    "url": "blob:https://webvm.io/53815c6a-4509-4c80-a9db-1c81410b4800",
    "attached": true,
    "parentId": "643492EFEDAED0AB10E09949114D5FCC",
    "canAccessOpener": false,
    "parentFrameId": "643492EFEDAED0AB10E09949114D5FCC",
    "browserContextId": "EC64F8722AA33D98E3AAB7D4DF7F2BEC"
}
```

**Comment by connorjclark** [comment]
The root cause seems to be that a worker on this page has entered a busy loop, never yielding back to the browser. This means devtools can't get any acks from commands it sends to that worker.

This page reproduces the same issue:

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Lighthouse Worker Hang Repro</title>
</head>
<body>
    <h1>Lighthouse Worker Hang Repro</h1>
    <p>This page spawns a Web Worker that immediately enters an infinite, synchronous loop.</p>
    <p>Before the fix, running a Lighthouse report on this page would cause DevTools to hang during the cleanup phase because it would wait indefinitely for the frozen worker target to respond to a CDP <code>resume</code> command.</p>

    <script>
        // Create an inline worker that runs an infinite loop
        const workerCode = `
            console.log("Worker started. Entering infinite loop...");
            // Block the worker's event loop entirely
            while (true) {
                // busy wait
            }
        `;

        const blob = new Blob([workerCode], { type: 'application/javascript' });
        const workerUrl = URL.createObjectURL(blob);
        
        console.log("Spawning worker...");
        const worker = new Worker(workerUrl);
        
        // This will never be called
        worker.onmessage = (e) => {
            console.log("Message from worker:", e.data);
        };
    </script>
</body>
</html>
```

**Comment by connorjclark** [comment]
Fixing upstream: https://chromium-review.git.corp.google.com/c/devtools/devtools-frontend/+/7885189

**Comment by connorjclark** [comment]
Should be resolved in Devtools for Chrome 150; and for Node in next release.
